src/unsafe_string.rs

```diff
@@ -0,0 +1,404 @@
+    }
+}
+
+trait HeaderMut<'a>: ThinMutExt<'a, Header> {
+    fn str_ptr_mut(mut self) -> *mut u8 {
+        // Safety: pointers to the end of structs are allowed
+        unsafe { self.ptr_mut().add(1).cast() }
+    }
+}
+
+impl<'a, T: ThinRefExt<'a, Header>> HeaderRef<'a> for T {}
+impl<'a, T: ThinMutExt<'a, Header>> HeaderMut<'a> for T {}
+
+static mut STRING_CACHE: Option<HashSet<WeakIString>> = None;
+
+// Eagerly initialize the string cache during tests or when the
+// `ctor` feature is enabled.
+#[cfg(any(test, feature = "ctor"))]
+#[ctor::ctor]
+unsafe fn ctor_init_cache() {
+    if STRING_CACHE.is_none() {
+        STRING_CACHE = Some(HashSet::new());
+    };
+}
+
+#[doc(hidden)]
+pub fn init_cache() {
+    unsafe {
+        if STRING_CACHE.is_none() {
+            STRING_CACHE = Some(HashSet::new());
+        };
+    }
+}
+
+#[cfg(not(feature = "ctor"))]
+unsafe fn get_cache() -> &'static mut HashSet<WeakIString> {
+    init_cache();
+    STRING_CACHE.as_mut().unwrap()
+}
+
+#[cfg(feature = "ctor")]
+unsafe fn get_cache() -> &'static mut HashSet<WeakIString> {
+    STRING_CACHE.as_mut().unwrap()
+}
+
+struct WeakIString {
+    ptr: NonNull<Header>,
+}
+
+impl PartialEq for WeakIString {
+    fn eq(&self, other: &Self) -> bool {
+        **self == **other
+    }
+}
+impl Eq for WeakIString {}
+impl Hash for WeakIString {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        (**self).hash(state);
+    }
+}
+
+impl Deref for WeakIString {
+    type Target = str;
+    fn deref(&self) -> &str {
+        self.borrow()
+    }
+}
+
+impl Borrow<str> for WeakIString {
+    fn borrow(&self) -> &str {
+        self.header().str()
+    }
+}
+impl WeakIString {
+    fn header(&self) -> ThinMut<Header> {
+        // Safety: pointer is always valid
+        unsafe { ThinMut::new(self.ptr.as_ptr()) }
+    }
+    fn upgrade(&self) -> IString {
+        unsafe {
+            self.header().rc += 1;
+            IString(IValue::new_ptr(
+                self.ptr.as_ptr().cast::<u8>(),
+                TypeTag::StringOrNull,
+            ))
+        }
+    }
+}
+
+/// The `IString` type is an interned, immutable string, and is where this crate
+/// gets its name.
+///
+/// Cloning an `IString` is cheap, and it can be easily converted from `&str` or
+/// `String` types. Comparisons between `IString`s is a simple pointer
+/// comparison.
+///
+/// The memory backing an `IString` is reference counted, so that unlike many
+/// string interning libraries, memory is not leaked as new strings are interned.
+/// Interning uses `DashSet`, an implementation of a concurrent hash-set, allowing
+/// many strings to be interned concurrently without becoming a bottleneck.
+///
+/// Given the nature of `IString` it is better to intern a string once and reuse
+/// it, rather than continually convert from `&str` to `IString`.
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct IString(pub(crate) IValue);
```

Comment:
This type still implements `Send + Sync` which it should not.

Also, it is insufficient for just `IString` to not implement `Send`/`Sync`. All `IValue` types would need to be `!Send` and `!Sync` when the `thread_safe` feature is not enabled, because an `IString` may be contained within another `IValue`.